File: ios/src/RNArcGISMapViewManager/RNArcGISMapViewManager.m
//
//  RNArcGISMapViewManager.m
//  SampleArcGIS
//
//  Created by David Galindo on 1/31/19.
//  Copyright © 2019 David Galindo. All rights reserved.
//  This is the objc-Bridge that connects Swift to React Native
//
#import <React/RCTBridgeModule.h>
#import <React/RCTViewManager.h>
#import "React/RCTEventEmitter.h"

@interface RCT_EXTERN_MODULE(RNArcGISMapViewManager, RCTViewManager)
RCT_EXPORT_VIEW_PROPERTY(basemapUrl, NSString)
RCT_EXPORT_VIEW_PROPERTY(routeUrl, NSString)
RCT_EXPORT_VIEW_PROPERTY(initialMapCenter, NSArray)
RCT_EXPORT_VIEW_PROPERTY(recenterIfGraphicTapped, BOOL)
RCT_EXPORT_VIEW_PROPERTY(minZoom, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(maxZoom, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(rotationEnabled, BOOL)
RCT_EXPORT_VIEW_PROPERTY(onSingleTap, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMapDidLoad, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onOverlayWasModified, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onOverlayWasAdded, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onOverlayWasRemoved, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMapMoved, RCTDirectEventBlock)


// MARK: External method exports (these can be called from react via a reference)
RCT_EXTERN_METHOD(showCalloutViaManager:(nonnull NSNumber*)node args:(NSDictionary*)args)
RCT_EXTERN_METHOD(centerMapViaManager:(nonnull NSNumber*)node args:(NSArray*)args)
RCT_EXTERN_METHOD(addGraphicsOverlayViaManager:(nonnull NSNumber*) node args:(NSDictionary*)args)
RCT_EXTERN_METHOD(removeGraphicsOverlayViaManager:(nonnull NSNumber*) node args:(NSString*) args)
RCT_EXTERN_METHOD(addPointsToOverlayViaManager:(nonnull NSNumber*) node args:(NSDictionary*)args)
RCT_EXTERN_METHOD(removePointsFromOverlayViaManager:(nonnull NSNumber*) node args:(NSDictionary*)args)
RCT_EXTERN_METHOD(updatePointsInGraphicsOverlayViaManager:(nonnull NSNumber*) node args:(NSDictionary*)args)
RCT_EXTERN_METHOD(routeGraphicsOverlayViaManager:(nonnull NSNumber*) node args:(NSDictionary*)args)
RCT_EXTERN_METHOD(routeIsShowing:(nonnull NSNumber*) node callback: (RCTResponseSenderBlock*)callback)
RCT_EXTERN_METHOD(setRouteIsVisibleViaManager:(nonnull NSNumber*) node args:(BOOL*) args)
RCT_EXTERN_METHOD(getRouteIsVisibleViaManager:(nonnull NSNumber*) node args:(BOOL*) args)
RCT_EXTERN_METHOD(setLicenseKey:(nonnull NSString*)key)

RCT_EXTERN_METHOD(dispose:(nonnull NSNumber*) node)
@end

@interface RCT_EXTERN_MODULE(RNArcGISMapViewModule, RCTEventEmitter)
RCT_EXTERN_METHOD(sendIsRoutingChanged: (BOOL*)value)
@end

File: ios/src/RNArcGISMapViewManager/RNArcGISMapViewManager.swift
//
//  RNArcGISMapViewManager.swift
//  SampleArcGIS
//
//  Created by David Galindo on 1/31/19.
//  Copyright © 2019 David Galindo. All rights reserved.
//

import Foundation
import ArcGIS


@objc(RNArcGISMapViewManager)
public class RNArcGISMapViewManager: RCTViewManager {
    var agsMapView: RNArcGISMapView?
    
    override public func view() -> UIView! {
        if (agsMapView == nil) {
            agsMapView = RNArcGISMapView()
            agsMapView!.bridge = self.bridge
        }
        return agsMapView!
    }
    
    override public class func requiresMainQueueSetup() -> Bool {
        return true;
    }
    
    // MARK: Exposed Obj-C bridging functions
    @objc func showCalloutViaManager(_ node: NSNumber, args: NSDictionary) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.showCallout(args)
        }
    }
    
    @objc func centerMapViaManager(_ node: NSNumber, args: NSArray) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.centerMap(args)
        }
    }
    
    @objc func addGraphicsOverlayViaManager(_ node: NSNumber, args: NSDictionary) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.addGraphicsOverlay(args)
        }
    }
    
    @objc func addPointsToOverlayViaManager(_ node: NSNumber, args: NSDictionary) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.addPointsToGraphicsOverlay(args)
        }
    }
    
    @objc func removePointsFromOverlayViaManager(_ node: NSNumber, args: NSDictionary) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.removePointsFromGraphicsOverlay(args)
        }
    }
    
    @objc func removeGraphicsOverlayViaManager(_ node: NSNumber, args: NSString) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.removeGraphicsOverlay(args)
        }
    }
    
    @objc func updatePointsInGraphicsOverlayViaManager(_ node: NSNumber, args: NSDictionary) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.updatePointsInGraphicsOverlay(args)
        }
    }
    
    @objc func routeGraphicsOverlayViaManager(_ node: NSNumber, args: NSDictionary) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.routeGraphicsOverlay(args)
        }
    }
    
    @objc func setRouteIsVisibleViaManager(_ node: NSNumber, args: ObjCBool) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.setRouteIsVisible(args.boolValue)
        }
    }
    
    @objc func getRouteIsVisibleViaManager(_ node: NSNumber, args: @escaping RCTResponseSenderBlock) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(forReactTag: node) as! RNArcGISMapView
            component.getRouteIsVisible(args)
        }
    }
    
    @objc func dispose(_ node: NSNumber) {
        self.agsMapView?.graphicsOverlays.removeAllObjects()
        self.agsMapView?.map = nil
        self.agsMapView = nil
    }
    
    @objc func setLicenseKey(_ key: String) {
        do {
            try AGSArcGISRuntimeEnvironment.setLicenseKey(key)
        }
        catch let error as NSError {
            print("error: \(error)")
        }
    }
}

@objc(RNArcGISMapViewModule)
public class RNArcGISMapViewModule: RCTEventEmitter {
    
    // MARK: Event emitting to JS
    @objc func sendIsRoutingChanged(_ value: Bool) {
        sendEvent(withName: "isRoutingChanged", body: [value])
    }
    
    
    // MARK: Overrides
    
    override public func supportedEvents() -> [String]! {
        return ["isRoutingChanged"]
    }
    
    @objc override public static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override public func constantsToExport() -> [AnyHashable : Any]! {
        return [:]
    }
    
}
